modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -0,0 +1,1655 @@
+            methodWriter.checkCast(MethodWriter.getType(getterPainlessMethod.returnType));
+        }
+    }
+
+    @Override
+    public void visitLoadMapShortcut(LoadMapShortcutNode irLoadMapShortcutNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irLoadMapShortcutNode.getLocation());
+
+        PainlessMethod getterPainlessMethod = irLoadMapShortcutNode.getGetter();
+        methodWriter.invokeMethodCall(getterPainlessMethod);
+
+        if (getterPainlessMethod.returnType != getterPainlessMethod.javaMethod.getReturnType()) {
+            methodWriter.checkCast(MethodWriter.getType(getterPainlessMethod.returnType));
+        }
+    }
+
+    @Override
+    public void visitLoadFieldMember(LoadFieldMemberNode irLoadFieldMemberNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irLoadFieldMemberNode.getLocation());
+
+        if (irLoadFieldMemberNode.isStatic()) {
+            methodWriter.getStatic(CLASS_TYPE,
+                    irLoadFieldMemberNode.getName(), MethodWriter.getType(irLoadFieldMemberNode.getExpressionType()));
+        } else {
+            methodWriter.loadThis();
+            methodWriter.getField(CLASS_TYPE,
+                    irLoadFieldMemberNode.getName(), MethodWriter.getType(irLoadFieldMemberNode.getExpressionType()));
+        }
+    }
+
+    @Override
+    public void visitLoadBraceDef(LoadBraceDefNode irLoadBraceDefNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irLoadBraceDefNode.getLocation());
+        Type methodType = Type.getMethodType(
+                MethodWriter.getType(irLoadBraceDefNode.getExpressionType()),
+                MethodWriter.getType(def.class),
+                MethodWriter.getType(irLoadBraceDefNode.getIndexType()));
+        methodWriter.invokeDefCall("arrayLoad", methodType, DefBootstrap.ARRAY_LOAD);
+    }
+
+    @Override
+    public void visitLoadBrace(LoadBraceNode irLoadBraceNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irLoadBraceNode.getLocation());
+        methodWriter.arrayLoad(MethodWriter.getType(irLoadBraceNode.getExpressionType()));
+    }
+
+    @Override
+    public void visitStoreVariable(StoreVariableNode irStoreVariableNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+
+        visit(irStoreVariableNode.getChildNode(), writeScope);
+
+        Variable variable = writeScope.getVariable(irStoreVariableNode.getName());
+        methodWriter.visitVarInsn(variable.getAsmType().getOpcode(Opcodes.ISTORE), variable.getSlot());
+    }
+
+    @Override
+    public void visitStoreDotDef(StoreDotDefNode irStoreDotDefNode, WriteScope writeScope) {
```

Comment:
See how difficult it is to make these visits alphabetical.

Reply:
After giving this some thought I'm going to skip it for now as it would require moving things around in 5 different files - 3 concrete phases and base classes. It also aligns better with rules all the way through from grammar -> user tree -> ir tree.